### PR TITLE
[Merged by Bors] - feat: number of edges in a complete graph

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean [hunk 1]
```diff
@@ -113,6 +108,21 @@ theorem sum_degrees_eq_twice_card_edges : ∑ v, G.degree v = 2 * G.edgeFinset.c
   G.dart_card_eq_sum_degrees.symm.trans G.dart_card_eq_twice_card_edges
 #align simple_graph.sum_degrees_eq_twice_card_edges SimpleGraph.sum_degrees_eq_twice_card_edges
 
+/-- The complete graph on `n` vertices has `n.choose 2` edges. -/
+theorem top_card_edges_eq_card_choose_two :
```
Comment (eric-wieser) [inline Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:112]:
```suggestion
theorem card_edgeFinset_top_eq_card_choose_two :
```

Comment (eric-wieser) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:112]: Though maybe the naming is inconsistent in this area of mathlib, and @YaelDillies or @b-mehta knows better

Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:112]: I agree with your name, Eric.

File: Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean [hunk 2]
```diff
@@ -113,6 +108,21 @@ theorem sum_degrees_eq_twice_card_edges : ∑ v, G.degree v = 2 * G.edgeFinset.c
...
+    (⊤ : SimpleGraph V).edgeFinset.card = (Fintype.card V).choose 2 := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:113]: I think we should rather have this lemma
```suggestion
    (⊤ : SimpleGraph V).edgeFinset = {e | ¬ e.IsDiag} := by
```

Comment (eric-wieser) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:113]: Surely you mean
```suggestion
    (⊤ : SimpleGraph V).edgeFinset = Finset.univ.filter fun e => ¬e
    IsDiag := by
```

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:113]: Done (for both variations).

File: Mathlib/Combinatorics/SimpleGraph/Basic.lean [hunk 3]
```diff
@@ -539,6 +539,13 @@ theorem edgeSet_bot : (⊥ : SimpleGraph V).edgeSet = ∅ :=
   Sym2.fromRel_bot
 #align simple_graph.edge_set_bot SimpleGraph.edgeSet_bot
 
+@[simp]
+theorem edgeSet_top : (⊤ : SimpleGraph V).edgeSet = {e | ¬e.IsDiag} := by
+  ext f
+  constructor <;> intro hm
+  · simp [(⊤ : SimpleGraph V).not_isDiag_of_mem_edgeSet hm]
+  · obtain ⟨_, _⟩ := f; exact hm
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Basic.lean:547]: If you do induction on `f`, aesop should be able to handle it.

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/Basic.lean:547]: This is clearly impossible. `f`'s type is `Sym2 V` which is not an inductive type.

Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Basic.lean:547]: Do you know about [`Quotient.inductionOn`](https://leanprover-community.github.io/mathlib4_docs/Init/Core.html#Quotient.inductionOn)?

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/Basic.lean]: Yeah, now everything works.

File: Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:113]: I would hope the proof now follows in terms of some general lemma about the cardinality of `Sym2`

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean]: Done.

File: Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean [hunk 4]
```diff
@@ -38,9 +38,7 @@ simple graphs, sums, degree-sum formula, handshaking lemma
 -/
 
 
-open Finset
-
-open BigOperators
+open Finset BigOperators
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:41]:
```suggestion
open Finset
open scoped BigOperators
```

File: Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean [hunk 5]
```diff
@@ -52,9 +50,6 @@ section DegreeSum
 
 variable [Fintype V] [DecidableRel G.Adj]
 
--- Porting note: Changed to `Fintype (Sym2 V)` to match Combinatorics.SimpleGraph.Basic
-variable [Fintype (Sym2 V)]
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:56]: Why this change? A priori you're making the lemma strictly less general.

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:56]: It's redundant – everything still compiles _without_ it. So the following lemmas cannot have been made less general by the removal of this variable.

Comment (eric-wieser) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:56]: That's not necessarily true, unfortunately; though certainly I would have expected a docstring explaining whether this is intentional.

Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:56]: @kmill, what's your opinion here?

Comment (kmill) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:56]: @Parcly-Taxel The rule of thumb with Fintype and Decidable instances is that you want to be sure that you have every instance be its own argument, even if that means redundancy. That's to help with all the overlapping instances we tend to have with these classes (especially since sometimes it's intentional, for example with `classical` and `Decidable`).

However -- Lean 4 doesn't seem to play as well with overlapping instances as in Lean 3 (it wants to resynthesize every instance from scratch, but Lean 3 was happy to use instances it picked up from unification), so this rule of thumb seems like it needs tweaking. It sounds plausible to me that if one `[...]` argument can be derived from the others, then there's no difference between including it or excluding it, so excluding it makes sense.

If we care about maximal rewritability, then we should be using `{...}` arguments for instances that can be picked up from unification. That's not great, though, because that tends to mean you need two versions of rewrite lemmas, since different instances tend to be picked up in the forwards and backwards directions. I wonder if we can convince core Lean to add "`[{...}]`" arguments that give us the Lean 3 behavior?

In short, ideally we'd have the extra `Fintype (Sym2 V)` argument for smoother rewriting, but at the moment I'm not sure it actually does that. I could be wrong though.

Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:56]: Let's not remove it for the moment then. It'll be much harder to find which lemmas could be generalised by having redundant arguments if those lemmas do not already have those arguments.

Comment (kmill) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean:56]: That seems reasonable to me.

To give a concrete issue that having an extra argument is meant to solve: you can obtain a `Fintype (Sym2 V)` instance from both [`Sym2.instFintype`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Finset/Sym.html#Sym2.instFintype) and [`Quotient.fintype`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Fintype/Basic.html#Quotient.fintype)

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/DegreeSum.lean]: Reverted.

File: Mathlib/Combinatorics/SimpleGraph/Basic.lean [hunk 6]
```diff
@@ -967,6 +971,22 @@ theorem edgeSet_univ_card : (univ : Finset G.edgeSet).card = G.edgeFinset.card :
...
+variable [Fintype V] [DecidableEq V]
+
+@[simp]
+theorem edgeFinset_top : (⊤ : SimpleGraph V).edgeFinset = univ.filter fun e => ¬e.IsDiag := by
+  ext f; exact f.inductionOn (by simp)
+
+/-- The complete graph on `n` vertices has `n.choose 2` edges. -/
+theorem card_edgeFinset_top_eq_card_choose_two :
+    (⊤ : SimpleGraph V).edgeFinset.card = (Fintype.card V).choose 2 := by
+  simp only [Set.toFinset_card, edgeSet_top, Set.coe_setOf, ← Sym2.card_subtype_not_diag]
```
Comment (kmill) [inline Mathlib/Combinatorics/SimpleGraph/Basic.lean:983]: Could you switch this to `simp_rw` so it's clear that it's a sequence of rewrites?
```suggestion
  simp_rw [Set.toFinset_card, edgeSet_top, Set.coe_setOf, ← Sym2.card_subtype_not_diag]
```
(It could be `rw` if only `edgeSet_top` didn't give "motive not type correct")

File: Mathlib/Combinatorics/SimpleGraph/Basic.lean [hunk 7]
```diff
@@ -539,6 +539,10 @@ theorem edgeSet_bot : (⊥ : SimpleGraph V).edgeSet = ∅ :=
   Sym2.fromRel_bot
 #align simple_graph.edge_set_bot SimpleGraph.edgeSet_bot
 
+@[simp]
+theorem edgeSet_top : (⊤ : SimpleGraph V).edgeSet = {e | ¬e.IsDiag} := by
+  ext f; exact f.inductionOn (by simp)
+
```
Comment (kmill) [inline Mathlib/Combinatorics/SimpleGraph/Basic.lean:545]: Could you add this additional theorem here?
```
@[simp]
theorem edgeSet_subset_setOf_not_isDiag : G.edgeSet ⊆ {e | ¬e.IsDiag} :=
  fun _ h => (Sym2.fromRel_irreflexive (sym := G.symm)).mp G.loopless h
```

File: Mathlib/Combinatorics/SimpleGraph/Basic.lean [hunk 8]
```diff
@@ -967,6 +971,22 @@ theorem edgeSet_univ_card : (univ : Finset G.edgeSet).card = G.edgeFinset.card :
...
+/-- The complete graph on `n` vertices has `n.choose 2` edges. -/
+theorem card_edgeFinset_top_eq_card_choose_two :
+    (⊤ : SimpleGraph V).edgeFinset.card = (Fintype.card V).choose 2 := by
+  simp only [Set.toFinset_card, edgeSet_top, Set.coe_setOf, ← Sym2.card_subtype_not_diag]
+
+/-- Any graph on `n` vertices has at most `n.choose 2` edges. -/
+theorem card_edgeFinset_le_card_choose_two : G.edgeFinset.card ≤ (Fintype.card V).choose 2 := by
+  apply (card_le_of_subset _).trans_eq card_edgeFinset_top_eq_card_choose_two
+  simp only [Set.subset_toFinset, Set.coe_toFinset, edgeSet_subset_edgeSet, le_top]
```
Comment (kmill) [inline Mathlib/Combinatorics/SimpleGraph/Basic.lean:988]: Here's a more conceptual proof. (I golfed this from a few `apply`s that show the steps better, but at least you can see the expected types in the info view to follow along).
```suggestion
  rw [← card_edgeFinset_top_eq_card_choose_two]
  exact card_le_of_subset (edgeFinset_mono le_top)
```

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/Basic.lean]: Done.

**Comment by kmill** [comment]
Thanks!

bors r+
